[Phase 1] Refactor 0x05h-Testing-Platform-Interaction.md (@NowSecure) (#2286)
OWASP/owasp-mastg: Document/0x05h-Testing-Platform-Interaction.md

This allows a common capability-style model where user interaction drives ad-hoc granting of fine-grained permission. This can be a key facility for reducing the permissions needed by apps to only those directly related to their behavior. Without this model in place malicious users may access other member's email attachments or harvest contact lists for future use via unprotected URIs. In the manifest the [`android:grantUriPermissions`](https://developer.android.com/guide/topics/manifest/provider-element#gprmsn "android:grantUriPermissions") attribute or the node help restrict the URIs.

Here you can find more information about APIs related to URI Permissions:

- [grantUriPermission](https://developer.android.com/reference/android/content/Context.html#grantUriPermission%28java.lang.String,%20android.net.Uri,%20int%29 "grantUriPermission")
- [revokeUriPermission](https://developer.android.com/reference/android/content/Context#revokeUriPermission%28android.net.Uri,%20int%29 "revokeUriPermission")
- [checkUriPermission](https://developer.android.com/reference/android/content/Context#checkUriPermission%28android.net.Uri,%20int,%20int,%20int%29 "checkUriPermission")

> ##### Custom Permissions

Reviewer: ```suggestion
### Custom Permissions
```

We recommend using a reverse-domain annotation when registering a permission, as in the example above (e.g. `com.domain.application.permission`) in order to avoid collisions with other applications.

> ## Injection Flaws
> 
> [Injection Flaws](https://owasp.org/www-community/Injection_Flaws "Injection Flaws") allow attackers to leverage the interconnection of applications and pass malicious code between the affected application and either another system (such as a server) or another connected client.
> 

Reviewer: Let's remove this one since we already have a more complete version in 0x04h-Testing-Code-Quality.md. I'll suggest a proper link below

```suggestion
```

> ### Deep Links
> 
> Android apps can expose functionality through deep links (which are a part of Intents). They can expose functionality to:
> 
> - other apps (via deep links or other IPC mechanisms, such as Intents or BroadcastReceivers).
> - the user (via the user interface).
> 
> None of the input from these sources can be trusted; it must be validated and/or sanitized. Validation ensures processing of data that the app is expecting only. If validation is not enforced, any input can be sent to the app, which may allow an attacker or malicious app to exploit app functionality.
> 

Reviewer: We can also remove this since it's already below.
```suggestion
```

> ## Fragment Injection

Reviewer: Let's move this back to the test since this is a very specific risk (and in the near future we'll actually move to the CODE category).

This abstract class parses the extra data fields of an Intent, in particular, the `PreferenceActivity.EXTRA_SHOW_FRAGMENT(:android:show_fragment)` and `Preference Activity.EXTRA_SHOW_FRAGMENT_ARGUMENTS(:android:show_fragment_arguments)` fields.

The first field is expected to contain the `Fragment` class name, and the second one is expected to contain the input bundle passed to the `Fragment`.

Because the `PreferenceActivity` uses reflection to load the fragment, an arbitrary class may be loaded inside the package or the Android SDK. The loaded class runs in the context of the application that exports this activity.

With this vulnerability, an attacker can call fragments inside the target application or run the code present in other classes' constructors. Any class that's passed in the Intent and does not extend the Fragment class will cause a `java.lang.CastException`, but the empty constructor will be executed before the exception is thrown, allowing the code present in the class constructor run.

> ## URL Loading in WebViews

Reviewer: Let's create a WebViews section and move everything about WebViews here.
```suggestion
### WebViews

#### URL Loading in WebViews
```

We recommend using a reverse-domain annotation when registering a permission, as in the example above (e.g. `com.domain.application.permission`) in order to avoid collisions with other applications.

### Fragment Injection

> Android SDK offers developers a way to present a [`Preferences activity`](https://developer.android.com/reference/android/preference/PreferenceActivity.html "Preference Activity") to users, allowing the developers to extend and adapt this abstract class.

Reviewer: ```suggestion
Android SDK offers developers a way to present a [`PreferencesActivity`](https://developer.android.com/reference/android/preference/PreferenceActivity.html "Preference Activity") to users, allowing the developers to extend and adapt this abstract class.
```

### Overview

> In order to test for [injection flaws](#injection-flaws "injection flaws") the following portions of the source code should be checked for app functionality that has been exposed:

Reviewer: ```suggestion
In order to test for [injection flaws](0x04h-Testing-Code-Quality.md#injection-flaws-mstg-arch-2-and-mstg-platform-2 "Injection Flaws") the following portions of the source code should be checked for app functionality that has been exposed:
```

> - Deep Links. Check the test case ["Testing Deep Links"](#testing-deep-links-mstg-platform-3) as well for further test scenarios.
> - IPC Mechanisms (Intents, Binders, Android Shared Memory, or BroadcastReceivers). Check the test case ["Testing for Sensitive Functionality Exposure Through IPC"](#testing-for-sensitive-functionality-exposure-through-ipc-mstg-platform-4) as well for further test scenarios.
> - User interface. Check the test case ["Testing for Overlay Attacks"](#testing-for-overlay-attacks-mstg-platform-9).

Reviewer: ```suggestion
- ["Testing Deep Links"](#testing-deep-links-mstg-platform-3)
- ["Testing for Sensitive Functionality Exposure Through IPC"](#testing-for-sensitive-functionality-exposure-through-ipc-mstg-platform-4)
- ["Testing for Overlay Attacks"](#testing-for-overlay-attacks-mstg-platform-9)
```

- File: `WebViewActivity.kt`
- Class: `com.mstg.deeplinkdemo.WebViewActivity`
- Method: `onCreate`

> Sometimes you can even take advantage of other applications that you know interact with your target app. You can reverse engineer the app, (e.g. to extract all strings and filter those which include the target deep links, `deeplinkdemo:///load.html` in the previous case), or use them as triggers, while hooking the app as previously discussed.

## Testing for Sensitive Functionality Exposure Through IPC (MSTG-PLATFORM-4)

### Overview

To test for [sensitive functionality exposure through IPC](#sensitive-functionality-exposure-through-ipc "Sensitive Functionality Exposure Through IPC") mechanisms we will identify a list of IPC mechanisms our app uses, then review the source code to see whether sensitive data is leaked when the mechanisms are used.

> ### Static Analysis

Reviewer: I can't send a suggestion for this. Please change 

> We start by looking at the AndroidManifest.xml, where all activities, services, and content providers included in the source code must be declared (otherwise the system won't recognize them and they won't run).

to

> We start by looking at the AndroidManifest.xml, where all activities, services, and content providers included in the app must be declared (otherwise the system won't recognize them and they won't run).

1. For the actual persistence: Use the techniques described in the data storage chapter.
2. For reflection-based approaches: Use Xposed to hook into the deserialization methods or add unprocessable information to the serialized objects to see how they are handled (e.g., whether the application crashes or extra information can be extracted by enriching the objects).

## Testing for Overlay Attacks (MSTG-PLATFORM-9)

### Overview

While testing for [overlay attacks](#overlay-attacks "Overlay Attacks") you can find some general guidelines about Android View security in the [Android Developer Documentation](https://developer.android.com/reference/android/view/View#security "View Security"), please be sure to read them carefully. For instance, the so-called _touch filtering_ is a common defense against tapjacking, which contributes to safeguarding users against these vulnerabilities, usually in combination with other techniques and considerations as we introduce in this section.

### Static Analysis

> To start your static analysis you can check the source code for the following methods and attributes (non-exhaustive list):

Reviewer: ```suggestion
To start your static analysis you can check the app for the following methods and attributes (non-exhaustive list):
```

Lastly, see if you can play with the version number of a man-in-the-middled app and see how the backend responds to this (and if it is recorded at all for instance).

## Testing WebViews Cleanup (MSTG-PLATFORM-10)

### Overview

When testing for [WebViews cleanup](#webviews-cleanup "WebViews Cleanup") there are a couple of areas where an app can delete WebView related data. You should inspect all related APIs and try to fully track data deletion.

- **WebView APIs**:
  - **Initialization**: an app might be initializing the WebView in a way to avoid storing certain information by using `setDomStorageEnabled`, `setAppCacheEnabled` or `setDatabaseEnabled` from [`android.webkit.WebSettings`](https://developer.android.com/reference/android/webkit/WebSettings "WebSettings"). The DOM Storage (for using the HTML5 local storage), Application Caches and Database Storage APIs are disabled by default, but apps might set these settings explicitly to "true".

>   - **Cache**: Android's WebView class offers the [`clearCache`](https://developer.android.com/reference/android/webkit/WebView#clearCache(boolean) "clearCache in WebViews") method which can be used to clear the cache for all WebViews used by the app. It receives a boolean input parameter (`includeDiskFiles`) which will wipe all stored resource including the RAM cache. However if it's set to false, it will only clear the RAM cache. Check the source code for usage of the `clearCache` method and verify its input parameter. Additionally, you may also check if the app is overriding `onRenderProcessUnresponsive` for the case when the WebView might become unresponsive, as the `clearCache` method might also be called from there.

Reviewer: ```suggestion
  - **Cache**: Android's WebView class offers the [`clearCache`](https://developer.android.com/reference/android/webkit/WebView#clearCache(boolean) "clearCache in WebViews") method which can be used to clear the cache for all WebViews used by the app. It receives a boolean input parameter (`includeDiskFiles`) which will wipe all stored resource including the RAM cache. However if it's set to false, it will only clear the RAM cache. Check the app for usage of the `clearCache` method and verify its input parameter. Additionally, you may also check if the app is overriding `onRenderProcessUnresponsive` for the case when the WebView might become unresponsive, as the `clearCache` method might also be called from there.
```
